src/Microsoft.PowerShell.Commands.Utility/commands/utility/sort-object.cs

```diff
@@ -71,19 +89,194 @@ protected override void EndProcessing()
             {
                 return;
             }
-            orderByProperty.OrderMatrix.Sort(orderByProperty.Comparer);
 
-            if (_unique)
+            // If -Unique was used, if -Top & -Bottom were not used, or of -Top or -Bottom would return all objects, sort
+            // all objects, remove duplicates if necessary, and identify the list of items to return
+            if (_unique || (Top == 0 && Bottom == 0) || Top >= orderByProperty.OrderMatrix.Count || Bottom >= orderByProperty.OrderMatrix.Count)
             {
-                RemoveDuplicates(orderByProperty);
+                orderByProperty.OrderMatrix.Sort(orderByProperty.Comparer);
+
+                if (_unique)
+                {
+                    RemoveDuplicates(orderByProperty);
+                }
+
+                var listToOutput = orderByProperty.OrderMatrix;
+
+                if (Top > 0 && Top < listToOutput.Count)
+                {
+                    // There may be a faster algorithm to sort while eliminating duplicates and retaining only
+                    // the top N records; for now though, this gets the job done
+                    listToOutput = listToOutput.GetRange(0, Top);
+                }
+                else if (Bottom > 0 && Bottom < listToOutput.Count)
+                {
+                    int lastItemIndex = listToOutput.Count - 1;
+                    listToOutput = listToOutput.GetRange(lastItemIndex - Bottom, lastItemIndex);
+                }
+
+                // write to output stream
+                foreach (var x in listToOutput)
+                {
+                    WriteObject(x.inputObject);
+                }
+            }
+            // Otherwise, use an indexed min-/max-heap to sort all objects
+            else
+            {
+                // The heap entry comparer will sort the items in the heap
+                var comparer = new IndexedMinMaxHeapComparer(orderByProperty.Comparer);
+
+                // Identify how many items will be in the heap
+                int indexedMinMaxHeapSize = Top > 0 ? Top : Bottom;
+
+                // A small list will store the data for the heap 
+                var indexedMinMaxHeap = new List<IndexedMinMaxHeapEntry>(indexedMinMaxHeapSize + 1);
```

Comment:
That's a good idea. I'll add that to my todo list, either for my blog or for PowerShell Magazine (maybe the latter -- it's been a while since I published an article there).